### PR TITLE
Set listeners h2 max streams to override nghttp2 client default of 100

### PR DESCRIPTION
**Pull request by kyessenov**
Reference issue: https://github.com/envoyproxy/envoy/issues/3076
Signed-off-by: Kuat Yessenov <kuat@google.com>

### CODE REVIEW

File: pilot/docker/envoy_pilot.yaml.tmpl [hunk 1]
```diff
@@ -135,6 +135,8 @@ static_resources:
               path: /dev/stdout
             name: envoy.file_access_log
           codec_type: HTTP2
+          http2_protocol_options:
+            max_concurrent_streams: 2147483646
```
Comment (PiotrSikora) [inline pilot/docker/envoy_pilot.yaml.tmpl]: That's overshooting it a it, the maximum one side can send is 2^30, i.e. 1,073,741,824.

Comment (kyessenov) [reply pilot/docker/envoy_pilot.yaml.tmpl]: Corrected. I was following the spec literally.

**Comment by geeknoid** [comment]
Isn't it dangerous to set this to effectively infinity? I thought there was a cost associated with each stream and so having a huge number there could potentially be used to DDoS the proxy.

**Comment by kyessenov** [comment]
@geeknoid We have a circuit breaker enabled on the client-side envoy, which should prevent it reaching the limit. I think it's set to 100,000 streams per client. The issue is mis-understanding the omitted value. The server assumes the unset value is infinity, which the client assumes it is 100. Making the server declare it to be infinity avoids the misunderstanding.

**Comment by geeknoid** [comment]
But for the server's own self-preservation perspective, wouldn't it be
safer to specify something smaller?

As we discussed yesterday, servers shouldn't trust clients to do the right
thing. So setting this number to infinity could let a client open an
infinite number of streams, which could kill the server. So maybe setting a
limit where (maxNumStreams*numBytesOfHeapNeededPerStream) is a reasonable
number for our sidecars would be preferable?

**Comment by kyessenov** [comment]
Currently, we permit the client to open an infinite number of connections which bypasses any limitations on streams per connection. I think once we do proper circuit breaking (limit both # connection and # streams per connection), we can reduce it to something reasonable.

**Comment by mandarjog** [comment]
From nghttp it looks like this setting can be overridden by the client, just the client is choosing not to override it. It does not appear to be a hard limit, and it is meant to keep initial requests from being rejected.

The documentation does not say that this is for dos protection.

I think this change is fine. @PiotrSikora

**Comment by geeknoid** [comment]
@louiscryan FYI

**Pull request merged**